### PR TITLE
Use owned data over refs for dto conversions

### LINKED ISSUE #563: Do DTO conversions really need a ref?

### PR DESCRIPTION
**Pull request by DaughterOfMars**
# Description of change

Do DTO conversions need to use refs? No. No they do not.

## Links to any relevant issues

- Closes #563

## Type of change

Choose a type of change, and delete any options that are not relevant.

- Breaking change (fix or feature that would cause existing functionality to not work as expected)

### CODE REVIEW

File: bindings/core/src/lib.rs [hunk 1]
```diff
@@ -49,7 +49,7 @@ pub struct WalletOptions {
 }
 
 impl WalletOptions {
-    pub async fn build_manager(&self) -> iota_sdk::wallet::Result<Wallet> {
+    pub async fn build_manager(self) -> iota_sdk::wallet::Result<Wallet> {
```
Comment (thibault-martinez) [inline bindings/core/src/lib.rs]: Just seeing this now but this should have been renamed a while ago. Can you rename to either `build_wallet` or simply `build` please? (+ changelog)

File: sdk/src/types/block/block.rs [hunk 2]
```diff
@@ -287,36 +287,40 @@ pub mod dto {
     }
 
     impl Block {
-        fn _try_from_dto(value: &BlockDto) -> Result<BlockBuilder, Error> {
```
Comment (thibault-martinez) [inline sdk/src/types/block/block.rs:290]: Any technical reason to remove the factorisation or you just didn't like it? For this one it's alright but some some it's a lot of repeated code

Comment (DaughterOfMars) [reply sdk/src/types/block/block.rs:290]: Same reason you had to `std::mem::take` before. The values are moved into the inner fn and cannot then be used afterward. We also can't do it the other way around, because we would have partially moved the value. I could have parameterized the inner fields but I felt that it wasn't worth it for a tiny bit of duplicated code. However, I am open to suggestions.

Comment (thibault-martinez) [reply sdk/src/types/block/block.rs:290]: Nah that's fine, was just wondering, thanks

File: sdk/src/client/secret/mod.rs [hunk 3]
```diff
@@ -154,12 +153,12 @@ impl FromStr for SecretManager {
     type Err = Error;
 
     fn from_str(s: &str) -> crate::client::Result<Self> {
-        Self::try_from(&serde_json::from_str::<SecretManagerDto>(s)?)
+        Self::try_from(serde_json::from_str::<SecretManagerDto>(s)?)
     }
 }
 
 /// DTO for secret manager types with required data.
-#[derive(Debug, Clone, Serialize, Deserialize, ZeroizeOnDrop)]
```
Comment (Thoralf-M) [inline sdk/src/client/secret/mod.rs:162]: Why did you remove `ZeroizeOnDrop`?

File: sdk/src/client/secret/mod.rs [hunk 4]
```diff
@@ -182,35 +181,37 @@ pub enum SecretManagerDto {
...
-    fn try_from(value: &SecretManagerDto) -> crate::client::Result<Self> {
+    fn try_from(value: SecretManagerDto) -> crate::client::Result<Self> {
         Ok(match value {
             #[cfg(feature = "stronghold")]
             SecretManagerDto::Stronghold(stronghold_dto) => {
                 let mut builder = StrongholdSecretManager::builder();
 
-                if let Some(password) = &stronghold_dto.password {
+                if let Some(password) = stronghold_dto.password {
                     // `SecretManagerDto` is `ZeroizeOnDrop` so it will take care of zeroizing the original.
```
Comment (Thoralf-M) [inline sdk/src/client/secret/mod.rs]: This comment can be removed then

File: sdk/src/types/block/payload/milestone/option/parameters.rs [hunk 5]
```diff
@@ -99,10 +99,10 @@ pub mod dto {
         }
     }
 
-    impl TryFrom<&ParametersMilestoneOptionDto> for ParametersMilestoneOption {
+    impl TryFrom<ParametersMilestoneOptionDto> for ParametersMilestoneOption {
         type Error = Error;
 
-        fn try_from(value: &ParametersMilestoneOptionDto) -> Result<Self, Self::Error> {
+        fn try_from(value: ParametersMilestoneOptionDto) -> Result<Self, Self::Error> {
```
Comment (Thoralf-M) [inline sdk/src/types/block/payload/milestone/option/parameters.rs:105]: Do we want to change it when it's not needed? Same for other changes

File: sdk/src/client/secret/mod.rs (hunk 3 above)
Comment (DaughterOfMars) [reply sdk/src/client/secret/mod.rs:162]: Because it means that you cannot move out of the type, and it doesn't actually make sense for this type to be zeroize, only the password.

File: sdk/src/types/block/payload/milestone/option/parameters.rs (hunk 5 above)
Comment (DaughterOfMars) [reply sdk/src/types/block/payload/milestone/option/parameters.rs:105]: What do you mean?

File: sdk/src/client/secret/mod.rs (hunk 3 above)
Comment (thibault-martinez) [reply sdk/src/client/secret/mod.rs:162]: What about the mnemonic and seed ?

Comment (DaughterOfMars) [reply sdk/src/client/secret/mod.rs:162]: Good point.

Comment (DaughterOfMars) [reply sdk/src/client/secret/mod.rs:162]: I had to impl a lot of changes that will probably be overwritten by #471 to fix it, just FYI

File: sdk/src/types/block/payload/milestone/option/parameters.rs (hunk 5 above)
Comment (Thoralf-M) [reply sdk/src/types/block/payload/milestone/option/parameters.rs:105]: What's the benefit of not using a ref here, when we don't clone anything anyways?

Comment (DaughterOfMars) [reply sdk/src/types/block/payload/milestone/option/parameters.rs:105]: Uniformity?

**Review by thibault-martinez — APPROVED** [review]

File: sdk/src/types/block/payload/milestone/option/parameters.rs (hunk 5 above)
Comment (Thoralf-M) [reply sdk/src/types/block/payload/milestone/option/parameters.rs:105]: In other places we change something to prevent a clone, but whatever

**Review by Thoralf-M — APPROVED** [review]
